lib/internal/modules/esm/loader.js

```diff
@@ -52,12 +52,59 @@ let hooksProxy;
  * @typedef {ArrayBuffer|TypedArray|string} ModuleSource
  */
 
+class HooksProxyLoaderDelegate {
+  /**
+   * Instantiate a module loader that uses user-provided custom loader hooks.
+   */
+  constructor() {
+    getHooksProxy();
+  }
+
+  /**
+   * Register some loader specifier.
+   * @param {string} originalSpecifier The specified URL path of the loader to
+   *                                   be registered.
+   * @param {string} parentURL The parent URL from where the loader will be
+   *                           registered if using it package name as specifier
+   * @returns {{ format: string, url: URL['href'] }}
+   */
+  register(originalSpecifier, parentURL) {
+    return hooksProxy.makeSyncRequest('register', originalSpecifier, parentURL);
+  }
+
+  /**
+   * Resolve the location of the module.
+   * @param {string} originalSpecifier The specified URL path of the module to
+   *                                   be resolved.
+   * @param {string} [parentURL] The URL path of the module's parent.
+   * @param {ImportAssertions} importAssertions Assertions from the import
+   *                                            statement or expression.
+   * @returns {{ format: string, url: URL['href'] }}
+   */
+  resolve(originalSpecifier, parentURL, importAssertions) {
+    if (importAssertions && importAssertions[kResolveSync]) {
+      return hooksProxy.makeSyncRequest('resolve', originalSpecifier, parentURL, importAssertions);
+    }
+    return hooksProxy.makeAsyncRequest('resolve', originalSpecifier, parentURL, importAssertions);
+  }
+
+  /**
+   * Provide source that is understood by one of Node's translators.
+   * @param {URL['href']} url The URL/path of the module to be loaded
+   * @param {object} [context] Metadata about the module
+   * @returns {Promise<{ format: ModuleFormat, source: ModuleSource }>}
+   */
+  load(url, context) {
+    return hooksProxy.makeAsyncRequest('load', url, context);
+  }
+}
 
 /**
- * This class covers the default case of an module loader instance where no custom user loaders are used.
- * The below CustomizedModuleLoader class extends this one to support custom user loader hooks.
+ * This class covers the base machinery of module loading. To add custom
+ * behaviour you can pass a delegate object and this object will be used
+ * to do the loading/resolving/registration process.
```

Comment:
Pass the delegate object on instantiation, you mean? And if you want to “convert” the class from uncustomized (default) to customized after instantiation, one should call `setDelegate`?

Could we get a name other than “delegate” for this? The word “delegate” is very generic, like “class” or “child.” Something to do with customization hooks?

Reply:
If you want to pick something I can find/replace the references before this lands 😄 I don't really have a better idea for a name that isn't already conflated with something else.